xdsl/transforms/experimental/dmp/stencil_global_to_local.py

```diff
@@ -235,100 +236,58 @@ def generate_memcpy(
 
     """
     assert ex.dim == 2, "Cannot handle non-2d case of memcpy yet!"
-    x0 = arith.Constant.from_int_and_width(ex.offset[0], builtin.IndexType())
+
+    idx = builtin.IndexType()
+
+    x0 = arith.Constant.from_int_and_width(ex.offset[0], idx)
     x0.result.name_hint = "x0"
-    y0 = arith.Constant.from_int_and_width(ex.offset[1], builtin.IndexType())
+    y0 = arith.Constant.from_int_and_width(ex.offset[1], idx)
     y0.result.name_hint = "y0"
-    x_len = arith.Constant.from_int_and_width(ex.size[0], builtin.IndexType())
+    x_len = arith.Constant.from_int_and_width(ex.size[0], idx)
     x_len.result.name_hint = "x_len"
-    y_len = arith.Constant.from_int_and_width(ex.size[1], builtin.IndexType())
+    y_len = arith.Constant.from_int_and_width(ex.size[1], idx)
     y_len.result.name_hint = "y_len"
-    cst0 = arith.Constant.from_int_and_width(0, builtin.IndexType())
-    cst1 = arith.Constant.from_int_and_width(1, builtin.IndexType())
-
-    # TODO: set to something like ex.size[1] < 8?
-    unroll_inner = False
-
-    # enable to get verbose information on what buffers are exchanged:
-    # print("Generating{} memcpy from buff[{}:{},{}:{}]{}temp[{}:{}]".format(
-    #    " unrolled" if unrolled else "",
-    #    ex.offset[0], ex.offset[0] + ex.size[0],
-    #    ex.offset[1], ex.offset[1] + ex.size[1],
-    #    '<-' if reverse else '->',
-    #    0, ex.elem_count
-    # ))
-
-    # only generate indices if we actually want to unroll
-    if unroll_inner:
-        indices = [
-            arith.Constant.from_int_and_width(i, builtin.IndexType())
-            for i in range(ex.offset[0], ex.offset[0] + ex.size[0])
-        ]
-    else:
-        indices = []
-
-    def loop_body_unrolled(i: SSAValue) -> Iterable[Operation]:
-        """
-        Generates last loop unrolled (not using scf.for)
-        """
-        dest_idx = arith.Muli(i, x_len)
-        y = arith.Addi(i, y0)
-        yield from (dest_idx, y)
-
-        for x in indices:
-            linearized_idx = arith.Addi(dest_idx, x)
-            if reverse:
-                load = memref.Load.get(dest, [linearized_idx])
-                store = memref.Store.get(load, source, [x, y])
-            else:
-                load = memref.Load.get(source, [x, y])
-                store = memref.Store.get(load, dest, [linearized_idx])
-            yield from (linearized_idx, load, store)
-        yield scf.Yield.get()
-
-    def loop_body_with_for(i: SSAValue) -> Iterable[Operation]:
+    cst0 = arith.Constant.from_int_and_width(0, idx)
+    cst1 = arith.Constant.from_int_and_width(1, idx)
+
+    @Builder.implicit_region([idx, idx])
```

Comment:
yay! clients!

Reply:
It's so wonderful to use! Thanks for adding it!